File: 07/gemini-3.0-pro/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error reading input.txt:", err)
		return
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if len(lines) == 0 {
		return
	}

	width := len(lines[0])
	height := len(lines)

	startCol := -1
	startRow := -1

	for r, line := range lines {
		idx := strings.Index(line, "S")
		if idx != -1 {
			startCol = idx
			startRow = r
			break
		}
	}

	if startCol == -1 {
		fmt.Println("Start not found")
		return
	}

	activeBeams := make(map[int]bool)
	activeBeams[startCol] = true
	totalSplits := 0

	for r := startRow + 1; r < height; r++ {
		nextBeams := make(map[int]bool)
		line := lines[r]

		for c := range activeBeams {
			if c < 0 || c >= width {
				continue
			}

			char := line[c]
			if char == '^' {
				totalSplits++
				if c-1 >= 0 {
					nextBeams[c-1] = true
				}
				if c+1 < width {
					nextBeams[c+1] = true
				}
			} else {
				nextBeams[c] = true
			}
		}
		activeBeams = nextBeams
		if len(activeBeams) == 0 {
			break
		}
	}

	fmt.Printf("Day 07 Part 1: %d\n", totalSplits)

	// Part 2
	memo := make(map[string]int64)
	var countPaths func(r, c int) int64
	countPaths = func(r, c int) int64 {
		if c < 0 || c >= width {
			return 1
		}

		key := fmt.Sprintf("%d,%d", r, c)
		if val, ok := memo[key]; ok {
			return val
		}

		for currR := r + 1; currR < height; currR++ {
			if lines[currR][c] == '^' {
				res := countPaths(currR, c-1) + countPaths(currR, c+1)
				memo[key] = res
				return res
			}
		}

		memo[key] = 1
		return 1
	}

	part2Ans := countPaths(startRow, startCol)
	fmt.Printf("Day 07 Part 2: %d\n", part2Ans)
}

File: 07/gemini-3.0-pro/rust/src/main.rs
use std::fs;
use std::collections::HashSet;

fn main() {
    let content = fs::read_to_string("input.txt").expect("Error reading input.txt");
    let lines: Vec<&str> = content.lines().collect();

    if lines.is_empty() {
        return;
    }

    let width = lines[0].len();
    let height = lines.len();

    let mut start_col = None;
    let mut start_row = 0;

    for (r, line) in lines.iter().enumerate() {
        if let Some(idx) = line.find('S') {
            start_col = Some(idx);
            start_row = r;
            break;
        }
    }

    let start_col = match start_col {
        Some(c) => c,
        None => {
            println!("Start not found");
            return;
        }
    };

    let mut active_beams: HashSet<usize> = HashSet::new();
    active_beams.insert(start_col);
    let mut total_splits = 0;

    for r in (start_row + 1)..height {
        let mut next_beams: HashSet<usize> = HashSet::new();
        let line_chars: Vec<char> = lines[r].chars().collect();

        for &c in &active_beams {
            if c >= width { continue; }

            let char = line_chars[c];
            if char == '^' {
                total_splits += 1;
                if c > 0 {
                    next_beams.insert(c - 1);
                }
                if c + 1 < width {
                    next_beams.insert(c + 1);
                }
            } else {
                next_beams.insert(c);
            }
        }
        active_beams = next_beams;
        if active_beams.is_empty() {
            break;
        }
    }

    println!("Day 07 Part 1: {}", total_splits);

    // Part 2
    use std::collections::HashMap;

    struct Solver<'a> {
        lines: &'a Vec<&'a str>,
        width: usize,
        height: usize,
        memo: HashMap<(usize, usize), u64>,
    }

    impl<'a> Solver<'a> {
        fn count_paths(&mut self, r: usize, c: usize) -> u64 {
            if c >= self.width {
                return 1;
            }

            if let Some(&res) = self.memo.get(&(r, c)) {
                return res;
            }

            for curr_r in (r + 1)..self.height {
                let char = self.lines[curr_r].as_bytes()[c] as char;
                if char == '^' {
                    let left_res = if c > 0 { self.count_paths(curr_r, c - 1) } else { 1 };
                    let right_res = self.count_paths(curr_r, c + 1);
                    let res = left_res + right_res;
                    self.memo.insert((r, c), res);
                    return res;
                }
            }

            self.memo.insert((r, c), 1);
            1
        }
    }

    let mut solver = Solver {
        lines: &lines,
        width,
        height,
        memo: HashMap::new(),
    };

    let part2_ans = solver.count_paths(start_row, start_col);
    println!("Day 07 Part 2: {}", part2_ans);
}
